Add LLC Quickstart Tutorial

ShivangiReja opened the pull request:
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/24493

ShivangiReja committed:
Add LLC guide start guide

ShivangiReja committed:
Quick fixes

ShivangiReja committed:
Quick fix

ShivangiReja committed:
Rename the doc

ShivangiReja committed:
Add introduction

ShivangiReja committed:
Add learn more

mikekistler commented on doc/LLC/LLC_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,181 @@
+
+In Visual Studio:
+
+- Rename the `Azure.Template.LLC` project to `Azure.<your-sdk-name>`.
```
For continuity with the "Repo Structure" documentation linked above, I think you should explain that `Azure.<your-sdk-name>` is the "package name".  And maybe change all the other instances of `Azure.<your-sdk-name>` to just `<package name>`.

ShivangiReja replied:
Thanks for the suggestion! I'll add the explanation that `Azure.<your-sdk-name>` is the "package name". I think `Azure.` is constant for track-2 libraries, so I added `Azure.<your-sdk-name>`. Please let me know if you still think that we should change `Azure.<your-sdk-name>` to `<package name>`.

ShivangiReja replied:
Thanks for the suggestion! I'll add the explanation that `Azure.<your-sdk-name>` is the "package name. In [this](https://github.com/Azure/azure-sdk-for-net/blob/27af1daac97a674ae723b218a41edbf0287abdbc/doc/LLC/LLC_Quickstart_Tutorial.md#3-rename-the-solution-and-projects) section in Learn more, I've added  [.NET Guidelines on Namespace Naming](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming), where it says

>✅ DO adhere to the following scheme when choosing a namespace: `Azure.<group>.<service>[.<feature>]`

I think `Azure.` is constant for track-2 libraries, so I added `Azure.<your-sdk-name>`. Please let me know if you still think that we should change `Azure.<your-sdk-name>` to something else.

mikekistler replied:
As long as there is a clear link to the "Repo Structure" doc I think we are good.  Thanks!

mikekistler commented on doc/LLC/LLC_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,181 @@
+
+- Add back `Azure.Template.LLC.Tests` csproj in your `Azure.<your-sdk-name>.sln`.
+- Rename the `Azure.Template.LLC.Tests` project to `Azure.<your-sdk-name>.Tests`.
+- Rename the `TemplateServiceTestEnvironment.cs` file to `<client-name>TestEnvironment.cs` and update the test environment. Before running live tests you will need to create live test resources, please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md) to learn more about how to manager resources and update test environment.
```
Here you use `<client-name>` without explaining what it is our how to find it.  Can you add some explanation here?

ShivangiReja replied:
Sure! Thanks for the suggestion.

mikekistler reviewed (commented):
This is great!  I love all the links to supporting material.

I left a few comments on very minor things I think could be improved.  Feel free to ignore, defer, or act on these as you see fit.

ShivangiReja committed:
Update doc/LLC/LLC_Quickstart_Tutorial.md

Co-authored-by: Mike Kistler <mikekistler@microsoft.com>

annelo-msft commented on doc/LLC/LLC_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,181 @@
+# LLC Quickstart Tutorial
```
Let's use this document to try out the new "data plane code generation" name, and see how it works in this context.

ShivangiReja replied:
Renamed LLC to "Dataplane Code Gen" in the doc. I'll update the template project links after we merge https://github.com/Azure/azure-sdk-for-net/pull/25390 PR.

ShivangiReja committed:
Mike's feedback

ShivangiReja committed:
Nit fixes

ShivangiReja committed:
Format

ShivangiReja committed:
Rename LLC to Dataplane Code Gen

ShivangiReja committed:
Fix format

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+# Dataplane Code Gen Quickstart Tutorial
```
```suggestion
# Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)
```

Let's write out data plane as two separate, unhyphenated words, as done here: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/control-plane-and-data-plane

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+# Dataplane Code Gen Quickstart Tutorial
+In this tutorial, we will walk through creating a new DataPlane Code Gen SDK based on [Azure SDK Design guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).
```
```suggestion
The purpose of the Azure SDK is to provide a unified developer experience across Azure services. In this tutorial, we will walk through creating a new Azure SDK Generated Client library for a data plane Azure service.  The output library will have a standardized API based on the  [.NET Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).  
```

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+# Dataplane Code Gen Quickstart Tutorial
+In this tutorial, we will walk through creating a new DataPlane Code Gen SDK based on [Azure SDK Design guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).
+
+We will use the [Dataplane Code Gen template](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-LLC) project as a reference. We have a template [swagger file](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-LLC/Azure.Template.LLC/src/swagger) containing a REST API definition based on [Azure REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNe ...
```
What if we did this?

```suggestion
We will use the [Azure SDK Generated Client template](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-dpg) project as a reference. We have a template [swagger file](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-LLC/Azure.Template.LLC/src/swagger) containing a REST API definition based on [Azure REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). We will use the swagger to generate SDK using [autorest.csharp](https://github.com/Azure/autorest.csharp).
```

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+For this tutorial, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. To create the project, we'll copy the [Azure.Template.DataplaneCodeGen](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-LLC) project for .NET.
+
+The `Azure.Template.DataplaneCodeGen` project is a great place to get started with Azure SDK .NET library development in general, as it contains a number of common patterns we use for tests, samples, documentation, generating API listings. It also has an example of a generated library for demonstrating Dataplane Code Gen APIs.
```
What about? 

```suggestion
The `Azure.Template.Generated` project is a great place to get started with Azure SDK .NET library development in general, as it contains a number of common patterns we use for tests, samples, documentation, generating API listings. It also has generated code demonstrating an example Data Plane Generated Client.
```

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+Open the `Azure.Template.DataplaneCodeGen.sln` file from the `sdk\<your-service-name>\Azure.Template.DataplaneCodeGen` directory.
+
+You'll notice that the `Azure.Template.DataplaneCodeGen` project already has generated code in it!  You're welcome to explore what's there -- it's been generated from the swagger file in the `swagger` folder. For this tutorial, however, we'll delete these generated files so we can start clean.
```
> in the `swagger` folder

provide a path starting from `sdk\...` for this?

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+In Visual Studio:
+
+- Rename the `Azure.Template.DataplaneCodeGen` project to `Azure.<your-sdk-name>`.
```
> to `Azure.<your-sdk-name>`.

Would it make sense to use the same naming convention as in the .NET guidelines?  `Azure.<group>.<service>`

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+**Note**: `Azure.<your-sdk-name>` is the Azure SDK package name.
+
+**Learn more:** see the [.NET Guidelines on Namespace Naming](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming) for more information on naming Azure SDK packages. This name diverges from approved Azure SDK namespaces, since it is intended to be used just for this tutorial.
```
It's worth noting that the package name/namespace will be something the architects will help standardize according to the guidelines.

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+In this section, we'll create a generated API layer built on Azure Core. We'll use a small swagger file containing a REST API definition for Azure Messaging WebPubSub.
+
+**Learn more:** the [autorest.csharp README](https://github.com/Azure/autorest.csharp#setup) has great samples showing how to customize the API in the generated code. Explore this further as you design APIs for your own service.
```
Instead of saying "customize" here, what about saying "add convenience APIs"?  Does that sound correct?

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
````diff
@@ -0,0 +1,185 @@
+> see https://aka.ms/autorest
+
+``` yaml
+input-file:
````
Some services use a different flag here and point to a readme.md file in the azure-rest-api-specs repo ... Is there documentation describing this that we could link out to?

ShivangiReja replied:
We don't have any docs about it. I added the `input-file` that partner team can use directly to add the swagger link for getting started with the new library.

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+- <swagger-file-link>
+namespace: <namespace>
+public-clients: true
+low-level-client: true
```
This will change to `--data-plane` ... I've created issue https://github.com/Azure/autorest.csharp/issues/1700 to track this.

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
`````diff
@@ -0,0 +1,185 @@
+```
+````
+
+**Learn more:** see the [.NET Guidelines for Namespaces](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming) to learn more about namespaces in the Azure SDK.
`````
Let's link to a reference for the readme.md instead of the namespace guidelines here.

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
````diff
@@ -0,0 +1,185 @@
+sdk\<your-service-name>\Azure.<your-sdk-name>\src> dotnet msbuild /t:GenerateCode
+```
+
+#### 3. Explore the Generated Code
````
I'm wondering if we can restructure this a bit to move the Explore part forward before the reader does the generation.  If we do, we can walk them through what's in the project before we have them start deleting/renaming things and refer to areas we know will be there.

ShivangiReja replied:
I kept this section here but also, I added the description at the top(second paragraph) of this section about the autorest.md and generated folder [here](https://github.com/Azure/azure-sdk-for-net/blob/9981a1f3c18dca95f186fe2adb8283e3c5caeb10/doc/Data%20Plane%20Code%20Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md#create-the-generated-layer).

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+## Snippets
+
+Snippets are the great way to reuse the sample code. Please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#updating-sample-snippets) to add snippets in your samples.
```
Would it be possible to call out that snippets allow us to verify that the code in our samples and READMEs is always up to date, and passes unit tests?

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
````diff
@@ -0,0 +1,185 @@
+eng\scripts\Export-API.ps1 webpubsub
+```
+
+## Readme
````
```suggestion
## README
```

annelo-msft commented on doc/Dataplane Code Gen/DataplaneCodeGen_Quickstart_Tutorial.md:
```diff
@@ -0,0 +1,185 @@
+
+## Readme
+
+README.md file instructions are listed in `Azure.<your-sdk-name>/README.md` file. Please add/update the README.md file as per your library.
```
Could we link to the bare template in a **Learn More** section here?  Could we also link to a good example of a README for one of the Track 2 libraries, maybe Key Vault?

annelo-msft reviewed (commented):
I added mostly nit-picky stuff and naming suggestions for new LLC names for here and in the template files.  The one larger item is with regard to the "Explore" section.  Would it be possible to add a discussion of what's in the Template file/the structure, etc?  We could do this either in once place in Getting Started or broken up in each of the sections where we explore before modifying them for the new service?

ShivangiReja committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into LLC_QuickstartGuide

ShivangiReja committed:
Feedback

ShivangiReja commented:
>The one larger item is with regard to the "Explore" section. Would it be possible to add a discussion of what's in the Template file/the structure, etc? We could do this either in once place in Getting Started or broken up in each of the sections where we explore before modifying them for the new service?

I added small description at the top of each section explaining what's in the template project. Some sections have already the description in it, I added for the rest of them.

ShivangiReja committed:
Nitfixes

chunyu3 commented on doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md:
```diff
@@ -0,0 +1,195 @@
+
+For this tutorial, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. To create the project, we'll copy the [Azure.Template.Generated](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-dpg) project for .NET.
+
+The `Azure.Template.Generated` project is a great place to get started with Azure SDK .NET library development in general, as it contains a number of common patterns we use for tests, samples, documentation, generating API listings. It also has generated code demonstrating an example Data Plane Generated Client.
```
In the template, src/properties/AssemblyInfo.cs, the NameSpace is not correct, shall we tell customer to correct it in this doc? And there are some Suppressions needed to added otherwise, build will fail. Shall we add those also?

ShivangiReja replied:
>In the src folder of template, there are a folder 'swagger' to store the swagger. Swagger should be in the azure-rest-api-specs, it should not be in the azure-sdk-for-net folder. It may be good not to add Swagger in the template

This is just a template project, in the template swagger I added all the CRUD operations which follows the REST API guidelines which gives the partner teams idea about the basic operations. Yes, we use the swagger from `azure-rest-api-specs` repo but they are too big to use in the template. Later in the doc I mentioned to remove the swagger folder, also I mentioned in the autorest.md section that add swagger to your service and gave an example.

annelo-msft commented on doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md:
````diff
@@ -0,0 +1,195 @@
+% git commit -m "initial commit for <service-name>"
+```
+
+### Create the generated layer
````
nit: (and below)
```suggestion
### Create the generated client
```

ShivangiReja commented on doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md:
```diff
@@ -0,0 +1,195 @@
+
+## Setup your repo
+
+- Fork and clone an [azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net) repo. Follow the instructions in the [.NET Contributing.md](https://github.com/Azure/azure-sdk-for-net/issues/12903) to fork and clone the `azure-sdk-for-net` repo.
```
This is intentional. I added the issue link because the instructions for fork and clone the repo is not added in the Contributing.md yet.

ShivangiReja commented on doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md:
```diff
@@ -0,0 +1,195 @@
+
+## Getting started
+
+For this tutorial, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. To create the project, we'll copy the [Azure.Template.Generated](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template-dpg) project for .NET.
```
We add the template project just like any other service SDK in azure-sdk-for-net repo. It gives the partner teams idea about the repo structure and where to add the sdk. 

I was also following the structure same as HLC template project. Here is an example.

ShivangiReja replied:
We add the template project just like any other service SDK in azure-sdk-for-net repo. It gives the partner teams idea about the repo structure and where to add the sdk. 

HLC template project also uses the same structure. [Here](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template) is an example.

ShivangiReja commented on doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md:
```diff
@@ -0,0 +1,195 @@
+
+You'll notice that the `Azure.Template.Generated` project already has generated code in it!  You're welcome to explore what's there -- it's been generated from the swagger file in the `sdk/template-dpg/Azure.Template.Generated/src/swagger` folder. For this tutorial, however, we'll delete these generated files so we can start clean.
+
+- Delete the Generated folder in the `Azure.Template.Generated` project.
```
I think it's good to see how a fully generated solution works, but interested to hear your perspectives - what would be a reason not to include these files in the template?

ShivangiReja replied:
I think it's good to see how a fully generated solution works, but interested to hear your perspectives - what would be a reason not to include these in the template?

ShivangiReja committed:
Add info about Updating Assembly Info file

ShivangiReja commented:
/check-enforcer override
